otherlibs/runtime_events/runtime_events.ml

```diff
@@ -156,25 +156,115 @@ module Timestamp = struct
+  type tag += UNK : tag
+
+  (* the data structure is primarily managed in C *)
+  type [@warning "-unused-field"] 'a t = {
+    id: int;
+    name: string;
+    typ: 'a Type.t;
+    tag: tag option;
+  }
+
+  external user_register : string -> tag option -> 'a Type.t -> 'a t
+                         = "caml_runtime_events_user_register"
+  external user_write : 'a t -> 'a -> unit = "caml_runtime_events_user_write"
+
+  let register name tag typ = user_register name (Some tag) typ
+
+  let write event value = user_write event value
+
+  let name ev = ev.name
+
+  let tag ev = Option.value ~default:UNK ev.tag
+end
+
 module Callbacks = struct
-    (* these record callbacks are only called from C code in the runtime
-       so we suppress the unused field warning *)
-    type[@warning "-unused-field"] t = {
-      runtime_begin: (int -> Timestamp.t -> runtime_phase -> unit) option;
-      runtime_end: (int -> Timestamp.t -> runtime_phase -> unit) option;
-      runtime_counter: (int -> Timestamp.t -> runtime_counter
-                        -> int -> unit) option;
-      alloc: (int -> Timestamp.t -> int array -> unit) option;
-      lifecycle: (int -> Timestamp.t -> lifecycle
-                  -> int option -> unit) option;
-      lost_events: (int -> int -> unit) option
-    }
-
-    let create ?runtime_begin ?runtime_end ?runtime_counter ?alloc ?lifecycle
-               ?lost_events () =
-      { runtime_begin; runtime_end; runtime_counter;
-          alloc; lifecycle; lost_events}
-  end
+
+  type 'a callback = int -> Timestamp.t -> 'a User.t -> 'a -> unit
+  (* Callbacks are bound to a specific event type *)
+  type any_callback = U : 'a callback -> any_callback
+
+  (* these record callbacks are only called from C code in the runtime
+      so we suppress the unused field warning *)
+  type[@warning "-unused-field"] t = {
+    runtime_begin: (int -> Timestamp.t -> runtime_phase -> unit) option;
+    runtime_end: (int -> Timestamp.t -> runtime_phase -> unit) option;
+    runtime_counter: (int -> Timestamp.t -> runtime_counter
+                      -> int -> unit) option;
+    alloc: (int -> Timestamp.t -> int array -> unit) option;
+    lifecycle: (int -> Timestamp.t -> lifecycle
+                -> int option -> unit) option;
+    lost_events: (int -> int -> unit) option;
+    (* user event callbacks is an array containing at each indice [i] a list
+        of functions to call when an event of type id [i] happen *)
+    user_events: any_callback list array;
+  }
+
+  let create ?runtime_begin ?runtime_end ?runtime_counter ?alloc ?lifecycle
+             ?lost_events () =
+    { runtime_begin; runtime_end; runtime_counter;
+        alloc; lifecycle; lost_events; user_events = Array.make 2 [] }
+
+  let add ty callback t =
+    let cur_max_id = Array.length t.user_events - 1 in
+    let id = Type.id ty in
+    let user_events =
+      if id > cur_max_id then
+        (* we need to resize the array. TODO maybe double the size ? *)
```

Comment:
The question remains: when the ID is out of bounds, should the size of the array be doubled (TODO) or should it be extended to the ID (current behavior) ?

Reply:
I think doubling might be safer. While we've got a minimum size of 256, we don't want be in a situation where eventually there are enough custom events that we end up doing lots of large array allocations in succession because we're continually resizing.